### PR TITLE
Add standalone / reusable dropdown button

### PR DESCRIPTION
**Pull request by jotaen**
So far the dropdown button was custom in the virtual media overlay. In order to reuse it, this PR introduces it as standalone component.

- By using a slot, the component for opening/closing the dropdown can be completely customised. So it can be any button variant, but also other elements. The arrow icon is therefore separate, so that its exact styling/position can be controlled from the outside too.
- In contrast to before (in the virtual media dialog) the global event listeners are now only active when the dropdown is open. Before, they were registered all the time, which is unnecessary.

Screenshot 2021-06-03 at 17 07 06

---
This change is [Reviewable](https://reviewable.io/reviews/mtlynch/tinypilot/721)

### CODE REVIEW

**Review by mtlynch — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mtlynch/tinypilot/721)* status: 0 of 3 files reviewed, 3 unresolved discussions (waiting on @jotaen)

___
*[app/templates/custom-elements/button-dropdown.html, line 58 at r1](https://reviewable.io/reviews/mtlynch/tinypilot/721#-MbHThX460Su1uV-ruOu:-MbHThX460Su1uV-ruOv:bvzpai6) ([raw file](https://github.com/mtlynch/tinypilot/blob/5ee99ed693586ae38f41a8c837b75c9321e298a3/app/templates/custom-elements/button-dropdown.html#L58)):*
> ```HTML
>             template.content.cloneNode(true)
>           );
>           this._handleCloseOnOutsideClick = this._handleCloseOnOutsideClick.bind(
> ```

Could we add parens here to avoid exceeding 80 columns?
___
*[app/templates/custom-elements/button-dropdown.html, line 76 at r1](https://reviewable.io/reviews/mtlynch/tinypilot/721#-MbHTpqG5mMXtCzfO9xQ:-MbHTpqG5mMXtCzfO9xR:b-wbm0rf) ([raw file](https://github.com/mtlynch/tinypilot/blob/5ee99ed693586ae38f41a8c837b75c9321e298a3/app/templates/custom-elements/button-dropdown.html#L76)):*
> ```HTML
> 
>         _toggleDropdown() {
>           const isShown = this.elements.dropdownItems.style.display === "block";
> ```

The abstraction here feels a little inconsistent since it's checking `style.display` directly here but manipulating it indirectly in `showDropdown` and `closeDropdown`. Could we make a `_isDropDownShowing` method?
___
*[app/templates/custom-elements/button-dropdown.html, line 98 at r1](https://reviewable.io/reviews/mtlynch/tinypilot/721#-MbHU9aH3l45Epr4CX_T:-MbHU9aH3l45Epr4CX_U:b-fy3psr) ([raw file](https://github.com/mtlynch/tinypilot/blob/5ee99ed693586ae38f41a8c837b75c9321e298a3/app/templates/custom-elements/button-dropdown.html#L98)):*
> ```HTML
>         _handleCloseOnOutsideClick(evt) {
>           const target = evt.composedPath()[0];
>           // Skip if the event target is the button of this component instance:
> ```

Can we rephrase this comment? I found it a little hard to parse. Can we put it more in terms of "If the user clicks on X" rather than "if the event target is..."?

**Review by jotaen — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mtlynch/tinypilot/721)* status: 0 of 3 files reviewed, 3 unresolved discussions (waiting on @mtlynch)

___
*[app/templates/custom-elements/button-dropdown.html, line 58 at r1](https://reviewable.io/reviews/mtlynch/tinypilot/721#-MbHThX460Su1uV-ruOu:-MbHZj6q8-Y423_mXu5m:bjple3w) ([raw file](https://github.com/mtlynch/tinypilot/blob/5ee99ed693586ae38f41a8c837b75c9321e298a3/app/templates/custom-elements/button-dropdown.html#L58)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

Could we add parens here to avoid exceeding 80 columns?
</blockquote></details>

I had actually formatted that differently (with a line break before `bind`), but Prettier always resets it to this. Not sure what you mean by the parens, though?
___
*[app/templates/custom-elements/button-dropdown.html, line 76 at r1](https://reviewable.io/reviews/mtlynch/tinypilot/721#-MbHTpqG5mMXtCzfO9xQ:-MbH_HJQ13vinal5hANY:b14qjgr) ([raw file](https://github.com/mtlynch/tinypilot/blob/5ee99ed693586ae38f41a8c837b75c9321e298a3/app/templates/custom-elements/button-dropdown.html#L76)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

The abstraction here feels a little inconsistent since it's checking `style.display` directly here but manipulating it indirectly in `showDropdown` and `closeDropdown`. Could we make a `_isDropDownShowing` method?
</blockquote></details>

:+1: Done.
___
*[app/templates/custom-elements/button-dropdown.html, line 98 at r1](https://reviewable.io/reviews/mtlynch/tinypilot/721#-MbHU9aH3l45Epr4CX_T:-MbHcJBGCMk3XVWbK-pP:b-896fix) ([raw file](https://github.com/mtlynch/tinypilot/blob/5ee99ed693586ae38f41a8c837b75c9321e298a3/app/templates/custom-elements/button-dropdown.html#L98)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

Can we rephrase this comment? I found it a little hard to parse. Can we put it more in terms of "If the user clicks on X" rather than "if the event target is..."?
</blockquote></details>

Done.

**Review by mtlynch — APPROVED** [review]
LGTM, thanks!

> *[Reviewable](https://reviewable.io/reviews/mtlynch/tinypilot/721#-:-MbHj8JfBXn8ARdBmj0K:b2c0zcb)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @jotaen and @mtlynch)

___
*[app/templates/custom-elements/button-dropdown.html, line 103 at r2](https://reviewable.io/reviews/mtlynch/tinypilot/721#-MbHimxUE5qx0TdC2y9Z:-MbHimxVBNM9G9Eu_gJY:b7u3p6e) ([raw file](https://github.com/mtlynch/tinypilot/blob/5920b45676d6245d28a6eb89ba9eb1ecc9efc753/app/templates/custom-elements/button-dropdown.html#L103)):*
> ```HTML
>           // Skip if the user has clicked on the main button of this component.
>           // In this case the click event is already handled by the event
>           // handler of the button.
> ```

I prefer to avoid passive voice, so consider  "In this case, the button's event handler processes the click event."

**Review by mtlynch — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mtlynch/tinypilot/721)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @jotaen)

___
*[app/templates/custom-elements/button-dropdown.html, line 58 at r1](https://reviewable.io/reviews/mtlynch/tinypilot/721#-MbHThX460Su1uV-ruOu:-MbHjGW5-CrGmwPWIi4c:b-d5onhq) ([raw file](https://github.com/mtlynch/tinypilot/blob/5ee99ed693586ae38f41a8c837b75c9321e298a3/app/templates/custom-elements/button-dropdown.html#L58)):*
<details><summary><i>Previously, jotaen (Jan Heuermann) wrote…</i></summary><blockquote>

I had actually formatted that differently (with a line break before `bind`), but Prettier always resets it to this. Not sure what you mean by the parens, though?
</blockquote></details>

Oh I mean could we add parentheses around it to allow prettier to reformat?

```javascript
= (this._handleCloseOnOutsideClick.bind(this));
```

Not sure if that will work, though.

**Review by jotaen — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mtlynch/tinypilot/721)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @mtlynch)

___
*[app/templates/custom-elements/button-dropdown.html, line 58 at r1](https://reviewable.io/reviews/mtlynch/tinypilot/721#-MbHThX460Su1uV-ruOu:-MbHy0792ArlHmc_wNut:b-78uc7m) ([raw file](https://github.com/mtlynch/tinypilot/blob/5ee99ed693586ae38f41a8c837b75c9321e298a3/app/templates/custom-elements/button-dropdown.html#L58)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

Oh I mean could we add parentheses around it to allow prettier to reformat?

```javascript
= (this._handleCloseOnOutsideClick.bind(this));
```

Not sure if that will work, though.
</blockquote></details>

That didn’t work unfortunately, I have tried multiple variations.

On a broader note: I think Prettier enforces its own rules for the most part and only allows few exceptions. As opposed to other formatters, I think the philosophy of Prettier is to treat [line-length more as a soft limit](https://prettier.io/docs/en/rationale.html#print-width). I guess if we wanted a hard cap at 80 chars, we’d need to combine Prettier with a Linter.

We also have line lengths > 80 chars in other JS-files, such as `app.js`.
___
*[app/templates/custom-elements/button-dropdown.html, line 103 at r2](https://reviewable.io/reviews/mtlynch/tinypilot/721#-MbHimxUE5qx0TdC2y9Z:-MbHxdWL3h_ggd-hcGp7:b-h4mlml) ([raw file](https://github.com/mtlynch/tinypilot/blob/5920b45676d6245d28a6eb89ba9eb1ecc9efc753/app/templates/custom-elements/button-dropdown.html#L103)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary><blockquote>

I prefer to avoid passive voice, so consider  "In this case, the button's event handler processes the click event."
</blockquote></details>

Ah, I should write a bot for the active/passive voice thing someday… don’t know why it’s so hard for me to get that right. Fixed.

**Comment by jotaen** [comment]
@mtlynch Could you merge into Pro please? 🙇‍♂️

**Comment by mtlynch** [comment]
@jotaen - Sure, done!
